### PR TITLE
Updated Repo Partial Delivery Function and Qualification

### PR DESCRIPTION
**Pull request by tomhealey-icma**
Cleaned up comments and verified build.

### CODE REVIEW

**Comment by tomhealey-icma** [comment]
All updates applied and build successful. Please review and approve.

**Review by chrisisla — CHANGES_REQUESTED** [review]
canceled = cancelled

File: rosetta-source/src/main/rosetta/event-common-func.rosetta [hunk 1]
```diff
@@ -1457,7 +1457,7 @@ func Qualify_OnDemandRateChange: <"The qualification of on an-demand rate change
 
 func Create_CancellationPrimitiveInstruction: <"Creates a primitive instruction for early cancellation.">
 	inputs:
-		tradeState TradeState (1..1) <"The original trade to be rolled.">
+		tradeState TradeState (1..1) <"The original trade to be canceled.">
```
Comment (chrisisla) [inline rosetta-source/src/main/rosetta/event-common-func.rosetta:1460]: canceled = cancelled

File: rosetta-source/src/main/rosetta/event-common-func.rosetta [hunk 2]
```diff
@@ -1824,7 +1830,81 @@ func Create_TerminationInstruction: <"Creates the relevant primitive instruction
 	set instruction -> quantityChange: <"Set primitive instruction to be only a quantity change instruction (with values set to zero).">
 		Create_QuantityChangeInstruction( [changePriceQuantity], QuantityChangeDirectionEnum->Replace, empty )
 
-// End of ICMA - Phase 2 Contribution
+func Create_PartialDeliveryPrimitiveInstruction: <"Creates the primitive instructions for a substitution of collateral by replacing the assetpayout of the trade.">
```
Comment (chrisisla) [inline rosetta-source/src/main/rosetta/event-common-func.rosetta:1833]: function comment is not correct

File: rosetta-source/src/main/rosetta/event-common-func.rosetta [hunk 3]
```diff
@@ -1824,7 +1830,81 @@ func Create_TerminationInstruction: <"Creates the relevant primitive instruction
...
+	inputs:
+		tradeState TradeState (1..1) <"The original trade to be closed.">
+		effectiveDate AdjustableOrRelativeDate (1..1) <"The date to close and open a new position.">
+		terminationDate AdjustableOrRelativeDate (1..1) <"The new termination date.">
+		deliveredPriceQuantity PriceQuantity (1..*) <"The price and quantity of the remaining trade. ">
```
Comment (chrisisla) [inline rosetta-source/src/main/rosetta/event-common-func.rosetta:1838]: think the comment for deliveryPriceQuantity should be "The price and quantity to set on the original trade" as this is the amount that has actually been delivered

File: rosetta-source/src/main/rosetta/event-common-func.rosetta [hunk 4]
```diff
@@ -1824,7 +1830,81 @@ func Create_TerminationInstruction: <"Creates the relevant primitive instruction
...
+		undeliveredPriceQuantity PriceQuantity (1..*) <"The price and quantity of the remaining trade. ">
+	output: 
+		instruction PrimitiveInstruction (1..1)
+
+	condition ContractualProduct: <"This function applies to contractual products only.">
+		tradeState -> trade -> tradableProduct -> product -> contractualProduct exists
+
+	set instruction -> split -> breakdown: <"Sets the first part of the split to be a termination instruction for the existing trade.">
```
Comment (chrisisla) [inline rosetta-source/src/main/rosetta/event-common-func.rosetta:1846]: this is not a termination instruction, its a quantity change, will it close the original trade?
looking at the ERCC guide and ISLA best practices it looks like the recommended action in both markets is that for a partial delivery the original trade is closed out and a new one created with the undelivered quantity. For both it also says that the original trade does not _need_ to be closed, but in this instance our best practices align.

File: rosetta-source/src/main/rosetta/event-common-func.rosetta [hunk 5]
```diff
@@ -1824,7 +1830,81 @@ func Create_TerminationInstruction: <"Creates the relevant primitive instruction
...
+		[ Create_PrimitiveInstruction(
+			empty, empty, empty, empty,
+						Create_QuantityChangeInstruction(
+						deliveredPriceQuantity, 
+						QuantityChangeDirectionEnum -> Replace, 
+						tradeState->trade->tradableProduct->tradeLot->lotIdentifier
+			),
+			empty, empty, empty, empty,	empty, empty, empty)]
+	
+	add instruction -> split -> breakdown: <"Sets the second part of the split to be a new contract with a new asset payout.">
```
Comment (chrisisla) [inline rosetta-source/src/main/rosetta/event-common-func.rosetta:1856]: its a new quantity not a new asset payout - and potentially new effective/term date based upon the termsChange in the PartialDeliveryInstructionTermsChange function. Might be good to update the comment accordingly

File: rosetta-source/src/main/rosetta/event-common-func.rosetta [hunk 6]
```diff
@@ -1824,7 +1830,81 @@ func Create_TerminationInstruction: <"Creates the relevant primitive instruction
...
+			empty, empty,
+			Create_PartialDeliveryInstructionTermsChange(
+				tradeState -> trade -> tradableProduct -> product -> contractualProduct,
+				effectiveDate, terminationDate
+				),
+			empty, empty, empty, empty
+			)
+		]
+	
+func Create_PartialDeliveryInstructionTermsChange:
```
Comment (chrisisla) [inline rosetta-source/src/main/rosetta/event-common-func.rosetta:1874]: add a function comment please

File: rosetta-source/src/main/rosetta/event-common-func.rosetta [hunk 7]
```diff
@@ -1824,7 +1830,81 @@ func Create_TerminationInstruction: <"Creates the relevant primitive instruction
...
+	inputs:
+		contractualProduct ContractualProduct (1..1) <"The contractual product to be modifier.">
```
Comment (chrisisla) [inline rosetta-source/src/main/rosetta/event-common-func.rosetta:1876]: modifier = modified

File: rosetta-source/src/main/rosetta/event-common-func.rosetta [hunk 8]
```diff
@@ -1824,7 +1830,81 @@ func Create_TerminationInstruction: <"Creates the relevant primitive instruction
...
+		termsChangeInstruction TermsChangeInstruction (1..1)
+
+	set termsChangeInstruction -> product -> contractualProduct:
+		contractualProduct
+	set termsChangeInstruction -> product -> contractualProduct -> economicTerms -> effectiveDate:
+		effectiveDate
+	set termsChangeInstruction -> product -> contractualProduct -> economicTerms -> terminationDate:
+		terminationDate
+
+func Qualify_PartialDelivery: <"Qualification of a partial delivery which constitutes a contract close and reopen with a difference quantity and termination date.">
```
Comment (chrisisla) [inline rosetta-source/src/main/rosetta/event-common-func.rosetta:1889]: difference = different

File: rosetta-source/src/main/rosetta/event-common-func.rosetta [hunk 9]
```diff
@@ -1824,7 +1830,81 @@ func Create_TerminationInstruction: <"Creates the relevant primitive instruction
...
+	output: is_event boolean (1..1)
+
+	alias beforeEconomicterms: businessEvent -> instruction only-element -> before -> trade -> tradableProduct -> product -> contractualProduct -> economicTerms
+	alias openEconomicTerms: FilterOpenTradeStates( businessEvent -> after ) only-element -> trade -> tradableProduct -> product -> contractualProduct -> economicTerms
+	alias closedTradeState: FilterClosedTradeStates( businessEvent -> after )
+
+	set is_event:
+		beforeEconomicterms exists and
+		openEconomicTerms exists and
+		closedTradeState count = 1 and
```
Comment (chrisisla) [inline rosetta-source/src/main/rosetta/event-common-func.rosetta:1902]: will the original trade be closed at this point? Will the first instruction in the split close it?

Comment (tomhealey-icma) [reply rosetta-source/src/main/rosetta/event-common-func.rosetta:1927]: Yes

File: rosetta-source/src/main/rosetta/event-common-func.rosetta (hunk 8 above)
Comment (tomhealey-icma) [reply rosetta-source/src/main/rosetta/event-common-func.rosetta:1914]: done

**Comment by tomhealey-icma** [comment]
I've replaced the three repo PRs with one new PR, Repo lifecycle additions (Partial Delivery, Reprice, Adjustment, Substitution) [https://github.com/finos/common-domain-model/pull/2472] (https://github.com/finos/common-domain-model/pull/2472).

**Pull request closed without merging**